File: ai_quality_checker/ai_quality_checker/report/ai_quality_inspection_log/ai_quality_inspection_log.js
// Copyright (c) 2024, Your Name and contributors
// For license information, please see license.txt

frappe.query_reports["AI Quality Inspection Log"] = {
    "filters": [
        {
            "fieldname": "supplier",
            "label": __("Supplier"),
            "fieldtype": "Link",
            "options": "Supplier"
        },
        {
            "fieldname": "status",
            "label": __("Status"),
            "fieldtype": "Select",
            "options": "\nPending\nPass\nFail"
        },
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1)
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today()
        }
    ],
    
    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        
        if (column.fieldname == "status") {
            if (value === "Pass") {
                value = "<span style='color: green; font-weight: bold;'>" + value + "</span>";
            } else if (value === "Fail") {
                value = "<span style='color: red; font-weight: bold;'>" + value + "</span>";
            } else if (value === "Pending") {
                value = "<span style='color: orange; font-weight: bold;'>" + value + "</span>";
            }
        }
        
        if (column.fieldname == "ai_confidence_score" && value) {
            const score = parseFloat(value);
            let color = "black";
            if (score >= 0.8) {
                color = "green";
            } else if (score >= 0.6) {
                color = "orange";
            } else {
                color = "red";
            }
            value = "<span style='color: " + color + "; font-weight: bold;'>" + score.toFixed(2) + "</span>";
        }
        
        return value;
    }
};

File: ai_quality_checker/ai_quality_checker/report/ai_quality_inspection_log/ai_quality_inspection_log.py
# Copyright (c) 2024, Your Name and contributors
# For license information, please see license.txt

import frappe
from frappe import _

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data

def get_columns():
    return [
        {
            "label": _("ID"),
            "fieldname": "name",
            "fieldtype": "Link",
            "options": "Material Quality Inspection",
            "width": 120
        },
        {
            "label": _("Item"),
            "fieldname": "item",
            "fieldtype": "Link",
            "options": "Item",
            "width": 150
        },
        {
            "label": _("Supplier"),
            "fieldname": "supplier",
            "fieldtype": "Link",
            "options": "Supplier",
            "width": 150
        },
        {
            "label": _("Inspection Date"),
            "fieldname": "inspection_date",
            "fieldtype": "Date",
            "width": 120
        },
        {
            "label": _("Status"),
            "fieldname": "status",
            "fieldtype": "Data",
            "width": 100
        },
        {
            "label": _("AI Confidence Score"),
            "fieldname": "ai_confidence_score",
            "fieldtype": "Float",
            "width": 130,
            "precision": 2
        },
        {
            "label": _("AI Remarks"),
            "fieldname": "ai_remarks",
            "fieldtype": "Data",
            "width": 300
        },
        {
            "label": _("Created On"),
            "fieldname": "creation",
            "fieldtype": "Datetime",
            "width": 150
        }
    ]

def get_data(filters):
    conditions = []
    values = []
    
    if filters.get("supplier"):
        conditions.append("supplier = %s")
        values.append(filters.get("supplier"))
    
    if filters.get("status"):
        conditions.append("status = %s")
        values.append(filters.get("status"))
    
    if filters.get("from_date"):
        conditions.append("inspection_date >= %s")
        values.append(filters.get("from_date"))
    
    if filters.get("to_date"):
        conditions.append("inspection_date <= %s")
        values.append(filters.get("to_date"))
    
    where_clause = ""
    if conditions:
        where_clause = "WHERE " + " AND ".join(conditions)
    
    query = f"""
        SELECT
            name,
            item,
            supplier,
            inspection_date,
            status,
            ai_confidence_score,
            ai_remarks,
            creation
        FROM
            `tabMaterial Quality Inspection`
        {where_clause}
        ORDER BY
            inspection_date DESC, creation DESC
    """
    
    return frappe.db.sql(query, tuple(values), as_dict=True)
